terraform/istio/install_asm.sh

```diff
@@ -0,0 +1,57 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#!/bin/bash
+
+# Install Istio on the single cluster for Stackdriver Sandbox
+
+# Download Istio
+
+echo "### "
+echo "### Begin install ASM control plane"
+echo "### "
+
+
+ASM_VERSION=1.12
+PROJECT_ID=$(gcloud config get-value project)
+CLUSTER_ZONE=$(gcloud container clusters list --filter name=cloud-ops-sandbox --format="json" | jq -r '.[0].zone')
+
+# Set the working directory to our current directory (/sandbox/terraform/istio)
+export SCRIPTPATH=$(dirname $(realpath $0))
+cd $SCRIPTPATH
+export WORK_DIR=$SCRIPTPATH
+
+echo "Downloading asmcli ${ASM_VERSION}..."
+curl https://storage.googleapis.com/csm-artifacts/asm/asmcli_${ASM_VERSION} > asmcli
+chmod +x asmcli
+
+echo "Installing ASM..."
+./asmcli install \
```

Comment:
Any reason why we can't use ASM Managed Control Plane (MCP) here? With MCP, `istiod` is out of the cluster which is saving some resources (CPU/Memory) usage as well as avoiding having to deal with ASM version upgrades. Could we consider this?

JFYI: If MCP is an option, this ASM TF module could help you with that: https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/modules/asm

Reply:
No reason. I'm not super knowledgeable about ASM, so I was just basing this off of what [I found in the public docs](https://cloud.google.com/service-mesh/docs/unified-install/install-anthos-service-mesh). I'll take a look at MCP.

I had seen that terraform module, but I was having trouble getting it working. I think we'd have to refactor more of our cluster creating configuration to use modules to get it to work properly. But I'm not super experienced with terrform either, so maybe it's just me.

We have an [open issue](https://github.com/GoogleCloudPlatform/cloud-ops-sandbox/issues/575) to refactor terraform into modules later, so I was thinking we would revisit using the ASM TF module as a part of that

Reply:
Yeah, MCP has plenty of advantages, when possible that's the recommendation to use it. In the case of the Cloud Ops Sandbox, here are the 2 main reasons why it's appealing:
- Get rid of the `istiod` component from the cluster to outside the clusters (less CPU/Memory usage in the cluster, more room for apps)
- The upgrade of ASM version is managed by Google, no need to upgrade the ASM/Istio version (`istiod` and the sidecar proxies).

To implement this with the new ASM TF module, it's very easy, see the example here: https://github.com/GoogleCloudPlatform/anthos-service-mesh-samples/blob/main/docs/terraform-asm-mcp/main.tf. No need to wrap the `amscli` bin yourself, very convenient and simplifying the installation experience.

Alternatively you could still leverage the ASM MCP by just updating what you do currently with the `asmcli` by adding the 2 parameters below.
```
./asmcli install
...
  --managed \
  --channel regular \
  --use_managed_cni
```
Here, like you mentioned it could be converted to the ASM TF module if you don't want to do it as part of this PR.

Hope that makes sense? We could also have a quick 1:1 chat internally, it will be my pleasure to connect with you :)

Reply:
Great, thanks for the detailed feedback, this is very helpful! I'm going to experiment a bit more with the terraform module, and will reach out to chat if needed.

Let me know if you notice anything else that seems off when you have time. We're hoping to merge this sometime next week